### DISCUSSION TITLE
Minimizing intersection between register() and additional logic on the same input tag.

### DISCUSSION THREAD

**Discussion by holem** [react-hook-form/react-hook-form#10004] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

The `register()` function returns 4 parameters that are supposed to be passed to `input` tag. Among them are 2 event handlers `onChange` and `onBlur`.

However, when I want to use these events for other logic (in my case, show tooltip `onFocus` and hide it by `onBlur`), I have to write an ugly workaround and the intended way of using `register()` does not work, because you can't just do `{...register("field")}` in input tag.

**Describe the solution you'd like**

One possible solution would be to remove the `onBlur` and `onChange` parameters from the return values of the `register()` function. Instead, it would be possible to subscribe to the relevant events inside the `register()` using the provided `ref`.

I understand that this improvement changes the API, but if people use `register()` in a standard way (i.e. like `<input {...register("name")}/>`), then everything will continue to work. Problems can arise only for those who, like me, also use `onBlur` and `onChange` events for additional logic in combination with handlers from `register()`.

**Describe alternatives you've considered**

To leave the existing API unchanged, it would be possible to provide a new register function with a new API which would not interfere with other logic on the same `input` tag.
```
const ref = registerField("name", options);
```
```
<input ref={registerField("name", options)} name="name" />
```
In this case you can also easily combine multiple refs on same `input` if you need it.

To support the use case when you spread params directly into `input` tag, the output of the function can contain 2 params - `ref` and `name` like so:
```
const {ref, name} = registerField("name", options);
```
```
<input {...registerField("name", options)} />
```
But I think this variant is redundant. Using only `ref` is much simpler, clearer and more versatile.

**Additional context**

Similar approach (subscribing to events by refs) is also used in other libraries for React, like react-dnd.

**Comment by bluebill1049** [comment]
> However, when I want to use these events for other logic (in my case, show tooltip onFocus and hide it by onBlur), I have to write an ugly workaround and the intended way of using register() does not work, because you can't just do {...register("field")} in input tag.

Could you provide more detail on the above?

Is this what you refer to?
`<input {...register('name')} onFocus={() => {}} />`

**Comment by Moshyfawn** [comment, marked as answer]
What "ugly workaround" are you referring to? RHF register options supports passing additional even handler logic as the second argument props like
```jsx
<input {...register("name", { onBlur: handleBlur })} />
```

**Comment by holem** [comment]
Thank you @bluebill1049 @Moshyfawn for your response and attention to this issue.

I ran into this problem when I designed the form and started adding tooltips to the fields. At that moment I have the following:

```javascript
<input {...register("name", options)} />
```

To add tooltip I need to subscribe to `onFocus` and `onBlur` events and out of habit I added them as follows:
```javascript
<input {...register("name", options)} onFocus={handleFocus} onBlur={handleBlur} />
```

And I was surprised why the behavior of the form changed even after `handleFocus` / `handleBlur` were empty functions. 
When I realized what the problem was, I made that "ugly workaround":
```javascript
const { onBlur, ...inputProps } = register("name", options);
const handleBlur = (e) => {
  // ...
  onBlur(e);
};
<input {...inputProps} onFocus={handleFocus} onBlur={handleBlur} />;
```
Thanks to @Moshyfawn I understand now that I could do instead like this:
```javascript
<input {...register("name", { onBlur: handleBlur })} onFocus={handleFocus} />
```

It's much better, but still not obvious and straightforward (at least to me). Because you need to keep in mind that to subscribe to `onBlur` and `onChange` events on `input` tag you should to do this only through `register`'s `options`, and all other events as usual. It can be a little confusing, and some may stumble over it (like me). And as for me there is no need to have this inconvenience if it can be easily solved (however, it requires making changes to the API), just like similar things are done in other successful libraries for React.

P.S. Anyway, gigantic thanks and huge respect to you and all people that working on this lib, you're amazing and help us a lot.

**Comment by bluebill1049** [comment]
Thanks for the update. Let's convert this into a a discussion instead and if others have similar concerns, we can look into improving, and thanks for the kind words to us 🙏

**Comment by Moshyfawn** [comment]
> And as for me there is no need to have this inconvenience

I'd argue that the spread syntax `register` today is a trade-off for the type-safety we got today. RHF _v6_ _(if I remember correctly)_ had the `register` passed to the `ref` prop it had it's own set of trade-offs.

While I see your point, I feel like it's brings a better DX to the table at the cost of this API "gotcha" you can look up / stumble upon once.

That said, I appreciate your take on the API and thank you for the kind words, @holem! ✨
